File: fetch2/src/main/java/com/tonyodev/fetch2/RequestData.java
package com.tonyodev.fetch2;

import android.support.annotation.NonNull;

import java.util.Map;

public final class RequestData {

    private final Status status;
    private final Error error;
    private final long downloadedBytes;
    private final long totalBytes;
    private final int progress;
    private final Request request;

    public RequestData(Request request, Status status, Error error,long downloadedBytes, long totalBytes, int progress) {
        this.request = request;
        this.status = status;
        this.error = error;
        this.downloadedBytes = downloadedBytes;
        this.totalBytes = totalBytes;
        this.progress = progress;
    }

    public long getId() {
        return request.getId();
    }

    @NonNull
    public String getUrl() {
        return request.getUrl();
    }

    @NonNull
    public String getAbsoluteFilePath() {
        return request.getAbsoluteFilePath();
    }

    @NonNull
    public Map<String, String> getHeaders() {
        return request.getHeaders();
    }

    @NonNull
    public Status getStatus() {
        return status;
    }

    @NonNull
    public Error getError() {
        return error;
    }

    public long getDownloadedBytes() {
        return downloadedBytes;
    }

    public long getTotalBytes() {
        return totalBytes;
    }

    public int getProgress() {
        return progress;
    }

    @NonNull
    public Request getRequest() {
        return request;
    }

    @NonNull
    public String getGroupId() {
        return request.getGroupId();
    }

    @Override
    public String toString() {
        return request.toString();
    }
}

File: fetch2/src/main/java/com/tonyodev/fetch2/RequestData.kt
package com.tonyodev.fetch2

import android.support.v4.util.ArrayMap

class RequestData(val url: String, val absoluteFilePath: String, status: Int,
                  error: Int, val downloadedBytes: Long, val totalBytes: Long, headers: MutableMap<String, String>, val groupId: String) {
    val status: Status
    val error: Error
    val progress: Int
    val headers: Map<String, String>
    val request: Request

    val id: Long
        get() = request.id

    init {
        var headers = headers

        if (url == null) {
            throw IllegalArgumentException("Url cannot be null")
        }

        if (absoluteFilePath == null) {
            throw IllegalArgumentException("AbsoluteFilePath cannot be null")
        }

        if (groupId == null) {
            throw IllegalArgumentException("groupId cannot be null")
        }

        if (headers == null) {
            headers = ArrayMap()
        }
        this.error = Error.valueOf(error)
        this.status = Status.valueOf(status)
        this.progress = DownloadHelper.calculateProgress(downloadedBytes, totalBytes)
        this.headers = headers
        this.request = Request(url, absoluteFilePath, headers)
    }

    override fun toString(): String {
        return request.toString()
    }
}
